p4-16/spec/P4-16-spec.mdk

```diff
@@ -5585,48 +5585,43 @@ control EPipe<T, IH, OH>(in IH inputHeaders,
                          in T fromIngress,
                          out OH outputHeaders,
                          out OutControl outCtrl);
-
 control Deparser<OH>(in OH outputHeaders, packet_out b);
 package Ingress<T, IH, OH>(Parser<IH> p,
-                           IPipe<_, IH, OH> map,
+                           IPipe<T, IH, OH> map,
                            Deparser<OH> d);
-package Egress<T, IH, OH>(Parser<IH> p, Port
-                          EPipe<_, IH, OH> map,
+package Egress<T, IH, OH>(Parser<IH> p,
+                          EPipe<T, IH, OH> map,
                           Deparser<OH> d);
-package Switch<T>( // Top-level switch contains two packages
-    // type types Ingress.IH and Egress.IH may be different
-    Ingress<T, _, _> ingress,
-    Egress<T, _, _> egress
-);
+package Switch<T>(Ingress<T, _, _> ingress, Egress<T, _, _> egress);
 ~ End P4Example
 
 Just from these declarations, even without reading a precise
 description of the target, the programmer can infer some useful
 information about the architecture of the described switch, as shown
 in Figure [#fig-switcharch]:
 
-- The switch contains two separate `package`s `Ingress` and `Egress`
+- The switch contains two separate `package`s `Ingress` and `Egress`.
 - The `Parser`, `IPipe`, and `Deparser` in the `Ingress` package
   are chained together in order. In addition, the `Ingress.IPipe` block has an
   input of type `Ingress.IH`, which is an output of the `Ingress.Parser`.
 - Similarly, the `Parser`, `EPipe`, and `Deparser` are
-  chained in the `Egress` package.
-- The `Ingress.IPipe` is connected to the `Egress.EPipe`,
-  because the first one outputs a value of type `T`, which is an
-  input to the second.
-- Note that the `Ingress` type `IH` and the `Egress` type `IH`
-  are independent. If we had wanted to show that they are the same
-  type, we would have instead declared `IH` and `OH` at the
-  switch level: `package Switch<IH, OH, T>`.
-
-Note that this architecture models a target switch that contains two
+  chained in the `Egress` package.  
+- The `Ingress.IPipe` is connected to the `Egress.EPipe`, because the
+  first outputs a value of type `T`, which is an input to the
+  second. Note that the the occurrences of the type variable `T` are
+  instantiated with the same type in `Switch`. In contrast, note that
+  the `Ingress` type `IH` and the `Egress` type `IH` may be
+  different. To force them to be the same type, we could instead
+  declare `IH` and `OH` at the switch level: `package Switch<IH, OH, T>`.
```

Comment:
Now I'm getting confused. I think both are legal programs but they mean different things... Let's just sort it out in person tomorrow?